autoextract/aio/client.py

```diff
@@ -34,6 +34,90 @@ class Result(List[Dict]):
     response_stats: Optional[List[ResponseStats]] = None
 
 
+class RequestProcessor:
+    """Help keeping track of query results and errors between retries.
+
+    After initializing your Request Processor,
+    you may use it for just a single or for multiple requests.
+
+    This class is especially useful because it stores
+    successful queries to avoid repeating them when retrying requests.
+    """
+
+    def __init__(self, query: Query, handle_retries: bool):
+        """Reset temporary data structures and initialize them"""
+        self._reset()
+        self.pending_queries = query_as_dict_list(query)
+        self._handle_retries = handle_retries
+        self._query_successes: List[Dict] = list()
+
+    def _reset(self):
+        """Clear temporary variables between retries"""
+        self.pending_queries: List[Dict] = list()
+        self._query_errors: List[Dict] = list()
+        self._retriable_query_exceptions: List[Dict] = list()
+
+    def _process_error(self, query_result, query_exception):
+        """Process Query-level error"""
+        self._query_errors.append(query_result)
+        self._retriable_query_exceptions.append(query_exception)
+
+        user_query = query_result["query"]["userQuery"]
+        self.pending_queries.append(user_query)
+
+    def get_latest_results(self):
+        """Get latest results (errors + successes).
+
+        This method could be used to retrieve results
+        when an exception is raised while processing results.
+        """
+        return self._query_successes + self._query_errors
+
+    def process_results(self, query_results):
+        """Process query results.
+
+        Return successful queries and also failed ones.
+
+        If `self._handle_exceptions` is True,
+        this method might raise a `QueryError` exception.
+
+        If multiple `QueryError` exceptions are parsed,
+        the one with the longest timeout is raised.
+
+        Successful requests are saved in `self._query_successes`
+        and they are kept between executions
+        while failures are saved in `self._query_errors`.
+
+        Queries saved in `self._query_errors` are moved to
+        `self.pending_queries` between executions.
+        You can use the first or the n-th result:
+
+            - You can get all queries successfully responded in the first try.
+            - You can get all queries successfully in the n-th try.
+            - You may stop with a partial number of successful queries.
+        """
+        self._reset()
+        for query_result in query_results:
+            if self._handle_retries and "error" in query_result:
+                query_exception = QueryError.from_query_result(query_result)
+                if query_exception.retriable:
+                    self._process_error(query_result, query_exception)
```

Comment:
Just not sure about the name, but I wouldn't bother too much about it.
Maybe `enqueue_errors`?

Reply:
Renamed